docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,494 @@
+- Scale down operations *could* be blocked based on the Health status
+  - for stacked etcd quorum must be considered
+  - for external etcd only availability of etcd should be considered.
+- Scale to 0 should be rejected for the initial support
+
+![controlplane-init-7](images/controlplane/controlplane-init-7.png)
+
+##### Delete of the entire ControlPlane (kubectl delete controlplane my-controlplane)
+
+- Completely removing the control plane and issuing a delete on the underlying machines.
+  - Deletion is equivalent to a scale to 1 followed by a deletion of a single replica Control Plane.
+
+##### Cluster upgrade (using create-swap-and-delete)
+
+- Triggered on any changes to Version, InfrastructureTemplate, or the KubeadmConfiguration used.
+  - Must be able to trigger an upgrade for the underlying machine image (e.g. nothing in the above list has changed but an upgrade is required)
+    - Possible implementations include (not necessary to hammer out exact implementation right now):
+      - Manually bumping a Generation field on the ControlPlane object
+      - Adding a field that is not the Generation field for users to manually bump to trigger an upgrade
+      - Adding a timestamp field, something like `nextUpgrade` which can be set to a specific time in the future (far future if no upgrades are desired) and an upgrade will run when that timestamp is passed. Good for scheduling upgrades/SLOs & quickly modifying to NOW or yesterday if an upgrade is required immediately. Would potentially update the field value to some interval (2 months from now?)
+- An upgrade will look like this:
+  - Serially go through each control plane replica not at the current config hash
+    - Check health of any replicas for the current config hash
+    - Uniquely mark the existing replica machine through setting an annotation and persisting the change (controlplane.cluster.x-k8s.io/selected-for-upgrade)
+    - Check cluster health
+    - Provision a new machine at the correct version
+    - Poll on health of replicas with the current config hash
+    - Poll on cluster health with the correct replica number
+    - Delete the marked controlplane machine
+- Upgrade of worker nodes is deferred to the MachineDeployment controller
+- Determining if a Machine is "up to date" will be done through the use of an annotation (controlplane.cluster.x-k8s.io/configuration-hash) that is placed on the Machine at creation time. The value of this annotation is generated by computing the Hash of the ControlPlaneSpec (minus the replicas field). This would allow triggering an Upgrade based on any changes to Version, InfrastructureTemplate, or the KubeadmConfiguration used.
+  - For example, a CVE is patched in containerd and updated images are available. Note that the Kubernetes version is not changed. To trigger an upgrade, the user updates the image in the InfrastructureTemplate (for an AWS cluster, the image is stored in InfrastructureTemplate.Spec.AMI.ID).
```

Comment:
Just curious how will the hash be calculated. Since infrastructureTemplate is an ObjectReference, so calculating hash  of ControlPlane will need to retrieve the actual object (AWSMachineTemplate) and include the Spec of it?

Reply:
@jzhoucliqr Since the Infrastructure templates are expected to be immutable after creation, we only need to rely on the reference. This is similar to how we treat the references in MachineDeployments today.

Reply:
got it. Thanks.